src/main/java/org/symphonyoss/symphony/messageml/elements/DatePicker.java

```diff
@@ -0,0 +1,321 @@
+    }
+    if(getAttribute(MIN_ATTR) != null){
+      assertDateFormat(MIN_ATTR, DateTimeFormatter.ISO_DATE);
+    }
+    if(getAttribute(MAX_ATTR) != null){
+      assertDateFormat(MAX_ATTR, DateTimeFormatter.ISO_DATE);
+    }
+    assertJsonDatesRange(DISABLED_DATE_ATTR);
+    assertJsonDatesRange(HIGHLIGHTED_DATE_ATTR);
+
+    if(getAttribute(FORMAT_ATTR) != null){
+      try {
+        DateTimeFormatter.ofPattern(getAttribute(FORMAT_ATTR));
+      } catch (IllegalArgumentException i){
+        throw new InvalidInputException(String.format("Attribute \"%s\" contains an invalid date format", FORMAT_ATTR));
+      }
+    }
+  }
+
+  @Override
+  public void asPresentationML(XmlPrintStream out,
+      MessageMLContext context) {
+    Map<String, String> presentationAttrs = buildDataPickerInputAttributes();
+    if(isSplittable()){
+      // open div + adding splittable elements
+      presentationAttrs.put(ID_ATTR, splittableAsPresentationML(out, context));
+      // render element
+      innerAsPresentationML(out, presentationAttrs);
+      // close div
+      out.closeElement();
+    } else {
+      innerAsPresentationML(out, presentationAttrs);
+    }
+  }
+
+  private void innerAsPresentationML(XmlPrintStream out, Map<String, String> presentationAttrs){
+    out.openElement(PRESENTATIONML_TAG, presentationAttrs);
+    out.closeElement();
+  }
+
+  private Map<String, String> buildDataPickerInputAttributes() {
+    Map<String, String> presentationAttrs = new LinkedHashMap<>();
+
+    presentationAttrs.put(NAME_ATTR, getAttribute(NAME_ATTR));
+
+    if(getAttribute(VALUE_ATTR) != null) {
+      presentationAttrs.put(VALUE_ATTR, getAttribute(VALUE_ATTR));
+    }
+    if(getAttribute(PLACEHOLDER_ATTR) != null) {
+      presentationAttrs.put(PLACEHOLDER_ATTR, getAttribute(PLACEHOLDER_ATTR));
+    }
+    if(getAttribute(MIN_ATTR) != null) {
+      presentationAttrs.put(MIN_ATTR, getAttribute(MIN_ATTR));
+    }
+    if(getAttribute(MAX_ATTR) != null) {
+      presentationAttrs.put(MAX_ATTR, getAttribute(MAX_ATTR));
+    }
+    if(getAttribute(DISABLED_DATE_ATTR) != null) {
+      presentationAttrs.put(DISABLED_DATE_PRESENTATION_ATTR, convertJsonDateToPresentationML(DISABLED_DATE_ATTR));
+    }
+    if(getAttribute(HIGHLIGHTED_DATE_ATTR) != null) {
+      presentationAttrs.put(HIGHLIGHTED_DATE_PRESENTATION_ATTR, convertJsonDateToPresentationML(HIGHLIGHTED_DATE_ATTR));
+    }
+    if(getAttribute(REQUIRED_ATTR) != null) {
+      presentationAttrs.put(REQUIRED_ATTR, getAttribute(REQUIRED_ATTR));
+    }
+    if(getAttribute(FORMAT_ATTR) != null) {
+      presentationAttrs.put(FORMAT_PRESENTATION_ATTR, getAttribute(FORMAT_ATTR));
+    }
+    return presentationAttrs;
+  }
+
+  /**
+   * The Json for PresentationML is different from MessageML
+   * It needs to be rewritten, by adding the type attribute, based on the content
+   *
+   * @param attributeName
+   * @return
+   */
+  private String convertJsonDateToPresentationML(String attributeName) {
```

Comment:
That's very complex :) Could you try to refactor the logic? Maybe using a Java POJO that you would then convert to JSON string

Reply:
Yes, I thought exactly the same thing after this first implementation, it is in progress...

Reply:
Refactored @symphony-thibault . Now the main class is much easier, even if the POJO used as a helper for the conversion is a little more than a POJO (it contains validation code too) and unfortunately there is this ugliness that I cannot fix: 

``java
result.replace("\"", "'")
``

...when converting with Jackson the POJO to a string. The problem is, single quotes are not officially supported in json format. Jackson can tolerate them for deserializing an object (if a specific mapper parameter is set!) but apparently no way to produce a json from an object with single quotes.